sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala

```diff
@@ -502,21 +500,20 @@ case class WholeStageCodegenRDD(@transient sc: SparkContext, var source: CodeAnd
 
   override def compute(split: Partition,
       context: TaskContext): Iterator[InternalRow] = {
-    val catcher = catching(classOf[ClassCastException])
     new Iterator[InternalRow] {
-      private[this] var i = computeInternal(split, context)
-
-      private[this] def replace() = i = {
-        logInfo(s"ClassCast Exception, hence recompiling")
-        CodeGenerator.invalidate(source)
-        computeInternal(split, context)
-      }
-
-      override def hasNext: Boolean = catcher.opt(i.hasNext).getOrElse {
-        replace(); hasNext
+      private[this] var iter = computeInternal(split, context)
+
+      override def hasNext: Boolean = try {
+        iter.hasNext
+      } catch {
+        case _: ClassCastException =>
+          logInfo(s"ClassCastException, hence recompiling")
+          CodeGenerator.invalidate(source)
+          iter = computeInternal(split, context)
```

Comment:
hasNext of the internal iterator is called, not the wrapping Iterator. The outer iterator has the logic of handling the CCE. If it got one, the exception handler will recompile the code again and call hasNext of inner iterator which can always fail if such genuine exception is there.